src/index.js

```diff
@@ -1,70 +1,699 @@
+                payload: { login, password }
+              })
+            }
+          })
+        }
+      }
+      if (loginError) {
+        this.bridge.emit('workerEvent', {
+          event: 'loginError',
+          payload: { msg: loginError.innerHTML }
+        })
+      }
+    })
+  }
+
+  onWorkerEvent({ event, payload }) {
+    this.log('info', 'onWorkerEvent starts')
+    if (event === 'loginSubmit') {
+      this.log('info', 'received loginSubmit, blocking user interactions')
+      this.blockWorkerInteractions()
+      const { login, password } = payload || {}
+      if (login && password) {
+        // On this website you could use your adherent number or your mail.
+        // We just follow de convention to save an "email"
+        // into the keyChain so there is no confusion when manipulating this credentials later
+        const email = login
+        this.store.userCredentials = { email, password }
+      }
+    } else if (event === 'loginError') {
+      this.log(
+        'info',
+        'received loginError, unblocking user interactions: ' + payload?.msg
+      )
+      this.unblockWorkerInteractions()
+    }
+  }
+
+  async navigateToLoginForm() {
+    this.log('info', '🤖 navigateToLoginForm')
+    await this.goto(baseUrl)
+    await Promise.all([
+      this.waitForElementInWorker('#Part_Vos-espaces_EspAss'),
+      this.waitForElementInWorker(
+        'a[href="https://agence.macif.fr/assurance/"]'
+      )
+    ])
+    await this.runInWorker('click', '#Part_Vos-espaces_EspAss')
+    await Promise.race([
+      this.waitForElementInWorker('#login'),
+      this.waitForElementInWorker('button[data-logout]'),
+      this.waitForElementInWorker('button', { includesText: 'Déconnexion' })
+    ])
+  }
+
+  async ensureAuthenticated({ account }) {
+    this.bridge.addEventListener('workerEvent', this.onWorkerEvent.bind(this))
+    this.log('info', '🤖 ensureAuthenticated')
+    if (!account) {
+      await this.ensureNotAuthenticated()
+    }
+    if (!(await this.isElementInWorker('#login'))) {
+      await this.navigateToLoginForm()
+    }
+    const authenticated = await this.runInWorker('checkAuthenticated')
+    if (!authenticated) {
+      this.log('info', 'Not authenticated')
+      const credentials = await this.getCredentials()
+      if (credentials) {
+        try {
+          // Full autoLogin is not possible on this website, there's a 2FA for each connection
+          // So we're only prefilling and sending the form to reach the 2FA page.
+          await this.prefillAndSendLoginForm()
+          this.log('info', 'Prefill successful, waiting for 2FA ...')
+        } catch {
+          this.log(
+            'info',
+            'Something went wrong with autoLogin, letting user log in'
+          )
+          await this.showLoginFormAndWaitForAuthentication()
+        }
+      } else {
+        await this.showLoginFormAndWaitForAuthentication()
+      }
+      await this.showLoginFormAndWaitForAuthentication()
```

Comment:
I don't  understand here why we run this.showLoginFormAndWaitForAuthentication() twice

Reply:
No specific reason, it's an oversight 👍

Reply:
You're right, just an oversight, i'll delete then second one 👍